### ISSUE TITLE
Crash when receiving "Access token expired (202)"

### ISSUE THREAD

**Issue by radu-tutueanu** [realm/realm-js#827]
The node-server-sdk crashes when receiving a "Access token expired (202)" message from the server.

Stacktrace
```
* thread #7: tid = 0x8ebbd, 0x000000010014928d node`v8::HandleScope::Initialize(v8::Isolate*) + 29, stop reason = EXC_BAD_ACCESS (code=1, address=0x6c08)
  * frame #0: 0x000000010014928d node`v8::HandleScope::Initialize(v8::Isolate*) + 29
    frame #1: 0x0000000104007ddd realm-node.node`realm::js::SyncClass<realm::node::Types>::populate_sync_config(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::Object>, realm::Realm::Config&)::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, realm::SyncConfig const&, std::__1::shared_ptr<realm::SyncSession>)::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, realm::SyncConfig const&, std::__1::shared_ptr<realm::SyncSession>) const + 63
    frame #2: 0x0000000104007d70 realm-node.node`void std::__1::__invoke_void_return_wrapper<void>::__call<realm::js::SyncClass<realm::node::Types>::populate_sync_config(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::Object>, realm::Realm::Config&)::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, realm::SyncConfig const&, std::__1::shared_ptr<realm::SyncSession>)&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, realm::SyncConfig const&, std::__1::shared_ptr<realm::SyncSession> >(realm::js::SyncClass<realm::node::Types>::populate_sync_config(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::Object>, realm::Realm::Config&)::'lambda'(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, realm::SyncConfig const&, std::__1::shared_ptr<realm::SyncSession>)&&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&&&, realm::SyncConfig const&&&, std::__1::shared_ptr<realm::SyncSession>&&) + 54
    frame #3: 0x00000001040e3932 realm-node.node`realm::_impl::sync_session_states::Active::access_token_expired(std::__1::unique_lock<std::__1::mutex>&, realm::SyncSession&) const + 126
    frame #4: 0x00000001040e1b81 realm-node.node`realm::SyncSession::handle_error(realm::SyncError) + 1217
    frame #5: 0x00000001040e50e5 realm-node.node`std::__1::__function::__func<realm::SyncSession::create_sync_session()::$_2, std::__1::allocator<realm::SyncSession::create_sync_session()::$_2>, void (std::__1::error_code, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator()(std::__1::error_code&&, bool&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 225
    frame #6: 0x0000000104331ac8 realm-node.node`void realm::sync::protocol::ClientProtocol::parse_message_received<(anonymous namespace)::Connection>((anonymous namespace)::Connection&, char const*, unsigned long) + 9816
    frame #7: 0x000000010432a047 realm-node.node`non-virtual thunk to (anonymous namespace)::Connection::websocket_binary_message_received(char const*, unsigned long) + 55
    frame #8: 0x00000001042e0d10 realm-node.node`(anonymous namespace)::WebSocket::frame_reader_loop() + 1712
    frame #9: 0x0000000104336cc4 realm-node.node`void realm::util::network::Service::AsyncOper::do_recycle_and_execute_helper<std::__1::function<void (std::__1::error_code, unsigned long)>, std::__1::error_code, unsigned long>(bool, bool&, std::__1::function<void (std::__1::error_code, unsigned long)>, std::__1::error_code, unsigned long) + 164
    frame #10: 0x0000000104336b0d realm-node.node`void realm::util::network::Service::AsyncOper::do_recycle_and_execute<std::__1::function<void (std::__1::error_code, unsigned long)>, std::__1::error_code&, unsigned long&>(bool, std::__1::function<void (std::__1::error_code, unsigned long)>&, std::__1::error_code&&&, unsigned long&&&) + 141
    frame #11: 0x00000001043386a1 realm-node.node`realm::util::network::Service::BasicStreamOps<realm::util::network::Socket>::BufferedReadOper<std::__1::function<void (std::__1::error_code, unsigned long)> >::recycle_and_execute() + 209
    frame #12: 0x00000001042d1235 realm-node.node`realm::util::network::Service::Impl::run() + 277
    frame #13: 0x000000010431d24c realm-node.node`realm::sync::Client::run() + 396
    frame #14: 0x00000001040dfd1e realm-node.node`std::__1::__thread_proxy<std::__1::tuple<realm::_impl::SyncClient::SyncClient(std::__1::unique_ptr<realm::util::Logger, std::__1::default_delete<realm::util::Logger> >, realm::sync::Client::ReconnectMode, bool)::'lambda'()> >(void*, void*) + 61
    frame #15: 0x00007fff9c872aab libsystem_pthread.dylib`_pthread_body + 180
    frame #16: 0x00007fff9c8729f7 libsystem_pthread.dylib`_pthread_start + 286
    frame #17: 0x00007fff9c8721fd libsystem_pthread.dylib`thread_start + 13
```

**Comment by radu-tutueanu** [comment]
cc @kristiandupont @fealebenpae

**Comment by radu-tutueanu** [comment]
I'll also check with the latest master - might be fixed by https://github.com/realm/realm-js/commit/5183ff9a8c176b56186c77ff1d57fcbcbe4b32b9

**Comment by radu-tutueanu** [comment]
@fealebenpae : Using the latest master, the client disconnects when the token expires but does not try reconnect:

```
Connection[1]: Connection closed due to error reported by server: Access token expired (202)
Connection[1]: Resolving '127.0.0.1:9080'
Connection[1]: Connecting to endpoint '127.0.0.1:9080' (1/1)
Connection[1]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:57292')
Connection[1]: Session[1]: Sending: BIND(server_path='/3dc672712d097862b038bfa3e4ff1b0d/test1', signed_user_token_size=573, need_file_ident_pair=0)
Connection[1]: Session[1]: Sending: IDENT(server_file_ident=8101741308083601252, client_file_ident=1, client_file_ident_secret=6620015922271569097, scan_server_version=11252, scan_client_version=11413, latest_server_version=11252, latest_server_session_ident=1897053923128638223)
Connection[1]: Received: ERROR(error_code=203, message_size=39, try_again=0, session_ident=1)
Connection[1]: Connection closed due to error reported by server: Bad user authentication (BIND, REFRESH) (203)
Closing Realm file: ./realm-object-server/3dc672712d097862b038bfa3e4ff1b0d/realm%3A%2F%2F127.0.0.1%3A9080%2F%7E%2Ftest1
```

**Comment by radu-tutueanu** [comment]
I did use the node-server-sdk

**Comment by fealebenpae** [comment]
@radu-tutueanu I believe this was fixed in 1.0.1. I know you've done some testing, but please post any findings you have here.

I created #840 to track the remainder of the work needed to proactively refresh the access token so as to not rely on reconnects.

**Comment by radu-tutueanu** [comment]
It was fixed, the only issue a little problematic is the fact that it takes 10-15 seconds to reconnect. I think that can be changed, for example Java reconnects immediately. #840 would fully fix the issue though.
